### PR TITLE
refactor(CMake): more consistent option management

### PR DESCRIPTION
**Pull request by coryan**
Use a single function to set compile features and options. The function
works for both our code, and code generated by Protobuf or gRPC.
Previously we used ad-hoc commands for the latter.

Part of the work for #14076

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14157)

### CODE REVIEW

File: cmake/GoogleCloudCppCommonOptions.cmake [hunk 1]
```diff
@@ -61,17 +87,18 @@ function (google_cloud_cpp_add_common_options target)
     if (GOOGLE_CLOUD_CPP_COMPILER_SUPPORTS_WCONVERSION)
         target_compile_options(${target} PRIVATE "-Wconversion")
     endif ()
+    # Disable a few warnings that are not very useful.
+    if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU"
+        AND "${CMAKE_CXX_COMPILER_VERSION}" VERSION_LESS 5.0)
+        # With GCC 4.x this warning is too noisy to be useful.
```
Comment (dbolduc) [inline cmake/GoogleCloudCppCommonOptions.cmake:93]: We don't support this compiler. Remove?

File: cmake/GoogleCloudCppCommonOptions.cmake [hunk 2]
```diff
@@ -29,29 +29,55 @@ check_cxx_compiler_flag(-Wno-sign-conversion
...
-    endif ()
-endfunction ()
+#[=======================================================================[.rst:
+google_cloud_cpp_add_common_options(target [NO_WARNINGS])
+---------------------------------------------------------
+
+Set the compilation options for the ``google-cloud-cpp`` targets.
 
+There are a number of compiler options and features we want to set on all our
+targets. This function sets them  During development we want to enable options
```
Comment (dbolduc) [inline cmake/GoogleCloudCppCommonOptions.cmake:39]: s/them*/them.$/

**Review by dbolduc — APPROVED** [review]

File: cmake/GoogleCloudCppCommonOptions.cmake (hunk 1 above)
Comment (coryan) [reply cmake/GoogleCloudCppCommonOptions.cmake]: Done

File: cmake/GoogleCloudCppCommonOptions.cmake (hunk 2 above)
Comment (coryan) [reply cmake/GoogleCloudCppCommonOptions.cmake]: Done

**Pull request merged**